<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Coexpression Analysis using WGCNA – Biología de Sistemas</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Biología de Sistemas</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./analisis_topo.html"> 
<span class="menu-text">Análisis y Topología (I)</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./analisis_topo_2.html"> 
<span class="menu-text">Análisis y Topología (II)</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./minimal_analysis.html"> 
<span class="menu-text">Minimal Análisis</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./clustering.html"> 
<span class="menu-text">Clustering</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./clustering_2.html"> 
<span class="menu-text">Clustering Enlaces</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./func_enric.html"> 
<span class="menu-text">Enriquecimiento Funcional</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./coexp.html" aria-current="page"> 
<span class="menu-text">Coexpression Analysis</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction-to-wgcna" id="toc-introduction-to-wgcna" class="nav-link active" data-scroll-target="#introduction-to-wgcna">Introduction to WGCNA</a>
  <ul class="collapse">
  <li><a href="#the-basics-of-wgcna" id="toc-the-basics-of-wgcna" class="nav-link" data-scroll-target="#the-basics-of-wgcna">The Basics of WGCNA</a></li>
  <li><a href="#module-identification" id="toc-module-identification" class="nav-link" data-scroll-target="#module-identification">Module Identification</a></li>
  <li><a href="#applications-in-systems-biology" id="toc-applications-in-systems-biology" class="nav-link" data-scroll-target="#applications-in-systems-biology">Applications in Systems Biology</a></li>
  </ul></li>
  <li><a href="#installation-of-required-packages" id="toc-installation-of-required-packages" class="nav-link" data-scroll-target="#installation-of-required-packages">Installation of Required Packages</a></li>
  <li><a href="#background-and-dataset-overview" id="toc-background-and-dataset-overview" class="nav-link" data-scroll-target="#background-and-dataset-overview">Background and Dataset Overview</a>
  <ul class="collapse">
  <li><a href="#data-files" id="toc-data-files" class="nav-link" data-scroll-target="#data-files">Data Files</a></li>
  </ul></li>
  <li><a href="#data-input-and-pre-processing" id="toc-data-input-and-pre-processing" class="nav-link" data-scroll-target="#data-input-and-pre-processing">Data Input and Pre-Processing</a></li>
  <li><a href="#identification-of-outlier-microarray-samples" id="toc-identification-of-outlier-microarray-samples" class="nav-link" data-scroll-target="#identification-of-outlier-microarray-samples">Identification of outlier microarray samples</a></li>
  <li><a href="#clinical-traits-integration" id="toc-clinical-traits-integration" class="nav-link" data-scroll-target="#clinical-traits-integration">Clinical Traits Integration</a></li>
  <li><a href="#network-construction-and-module-detection" id="toc-network-construction-and-module-detection" class="nav-link" data-scroll-target="#network-construction-and-module-detection">Network Construction and Module Detection</a></li>
  <li><a href="#correlating-modules-with-external-clinical-traits" id="toc-correlating-modules-with-external-clinical-traits" class="nav-link" data-scroll-target="#correlating-modules-with-external-clinical-traits">Correlating modules with external clinical traits</a>
  <ul class="collapse">
  <li><a href="#quantifying-moduletrait-associations" id="toc-quantifying-moduletrait-associations" class="nav-link" data-scroll-target="#quantifying-moduletrait-associations">Quantifying module–trait associations</a></li>
  <li><a href="#plot-module-trait-correlations" id="toc-plot-module-trait-correlations" class="nav-link" data-scroll-target="#plot-module-trait-correlations">Plot module-trait correlations</a></li>
  </ul></li>
  <li><a href="#gene-relationship-to-trait-and-important-modules-gene-significance-and-module-membership" id="toc-gene-relationship-to-trait-and-important-modules-gene-significance-and-module-membership" class="nav-link" data-scroll-target="#gene-relationship-to-trait-and-important-modules-gene-significance-and-module-membership">Gene relationship to trait and important modules: Gene Significance and Module Membership</a>
  <ul class="collapse">
  <li><a href="#why-is-module-membership-useful" id="toc-why-is-module-membership-useful" class="nav-link" data-scroll-target="#why-is-module-membership-useful">Why Is Module Membership Useful?</a></li>
  </ul></li>
  <li><a href="#summary-output-of-network-analysis-results" id="toc-summary-output-of-network-analysis-results" class="nav-link" data-scroll-target="#summary-output-of-network-analysis-results">Summary output of network analysis results</a></li>
  <li><a href="#now-your-task---select-all-genes-that-belong-to-the-brown-module-and-perform-ora-to-see-if-you-can-find-enrichment-for-any" id="toc-now-your-task---select-all-genes-that-belong-to-the-brown-module-and-perform-ora-to-see-if-you-can-find-enrichment-for-any" class="nav-link" data-scroll-target="#now-your-task---select-all-genes-that-belong-to-the-brown-module-and-perform-ora-to-see-if-you-can-find-enrichment-for-any">Now your task - select all genes that belong to the brown module and perform ORA to see if you can find enrichment for any</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Coexpression Analysis using WGCNA</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction-to-wgcna" class="level1">
<h1>Introduction to WGCNA</h1>
<p>Weighted Gene Co-expression Network Analysis (WGCNA) is a powerful bioinformatics method used to describe the correlation patterns among genes across multiple samples. The method identifies gene modules, summarizes them using module eigengenes, and correlates these with external traits or phenotypes.</p>
<section id="the-basics-of-wgcna" class="level2">
<h2 class="anchored" data-anchor-id="the-basics-of-wgcna">The Basics of WGCNA</h2>
<p>Gene co-expression networks are built by calculating pairwise correlations between gene expression profiles. To capture biological significance and robustness, WGCNA uses a <strong>soft-thresholding power</strong> to transform the correlation matrix into an adjacency matrix. This adjacency matrix <span class="math inline">\(a_{ij}\)</span> represents the connection strength between genes <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, defined as:</p>
<p><span class="math display">\[
a_{ij} = |cor(x_i, x_j)|^\beta
\]</span></p>
<p>Where: - <span class="math inline">\(cor(x_i, x_j)\)</span> is the correlation between gene <span class="math inline">\(i\)</span> and gene <span class="math inline">\(j\)</span>. - <span class="math inline">\(\beta\)</span> is the soft-thresholding power, chosen to approximate scale-free network topology.</p>
</section>
<section id="module-identification" class="level2">
<h2 class="anchored" data-anchor-id="module-identification">Module Identification</h2>
<p>Modules are groups of highly interconnected genes. They are identified using hierarchical clustering of the Topological Overlap Matrix (TOM), which measures the similarity of gene neighborhoods. The TOM is calculated as:</p>
<p><span class="math display">\[
TOM_{ij} = \frac{\sum_k (a_{ik} \cdot a_{jk}) + a_{ij}}{\min(k_i, k_j) + 1 - a_{ij}}
\]</span></p>
<p>Where:</p>
<p><span class="math inline">\(k_i, k_j\)</span> = The degree (sum of edge weights) of nodes <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, respectively.</p>
<p>Intuitively, the numerator combines the shared neighbors and the direct connections between <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>. The denominator normalizes this value based on the smaller degree of the two nodes (<span class="math inline">\(min(k_ik_j)\)</span>), ensuring comparability across the network.</p>
<p>TOM thus accounts for Shared Neighbors - Nodes that are not directly connected but share many neighbors will have a high TOM value. This captures indirect relationships, making TOM more robust to noise than simple pairwise correlations. It is also a weighted measure: TOM is based on weighted adjacency values, allowing it to reflect the strength of relationships rather than just their presence or absence.</p>
</section>
<section id="applications-in-systems-biology" class="level2">
<h2 class="anchored" data-anchor-id="applications-in-systems-biology">Applications in Systems Biology</h2>
<p>By identifying gene modules and their relationships with traits, WGCNA facilitates the discovery of key drivers and functional insights in systems biology. This tutorial applies WGCNA to liver expression data from female mice, focusing on identifying gene modules related to body weight.</p>
</section>
</section>
<section id="installation-of-required-packages" class="level1">
<h1>Installation of Required Packages</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install BiocManager</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Install WGCNA</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"WGCNA"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"WGCNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the required library</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(WGCNA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: dynamicTreeCut</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: fastcluster</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'fastcluster'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stats':

    hclust</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'WGCNA'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stats':

    cor</code></pre>
</div>
</div>
<hr>
</section>
<section id="background-and-dataset-overview" class="level1">
<h1>Background and Dataset Overview</h1>
<p>Obesity is a major public health concern in many developed countries. While some people appear to stay lean no matter what or how much they eat, others appear to be genetically predisposed to obesity. The genetic similarity between mouse and human makes the mouse a promising mammalian model system to study obesity.</p>
<p>The mouse strain C57BL/6J is susceptible to a variety of ailments related to atherosclerosis, diabetes, obesity, and heart disease to which the mouse strain C3H/HeJ is resistant. The offspring mice of this F2 mouse intercross are expected to show a significant spectrum of atherosclerosis and metabolic syndrome responses to a high-fat diet. A variety of physiological traits were measured, including mouse body weight, fat mass, insulin, glucose, free fatty-acid levels in the blood, and cholesterol fractions (HDL and LDL + VLDL). Excerpt From: Steve Horvath. “Weighted Network Analysis.”</p>
<p>We will use microarray data from the females in the F2 mouse intercross to examine the large-scale organization of the gene co-expression network in liver, and find gene modules associated with body weight.</p>
<section id="data-files" class="level3">
<h3 class="anchored" data-anchor-id="data-files">Data Files</h3>
<ul>
<li><strong>Expression data</strong>: <a href="LiverFemale3600.csv" class="external" target="_blank">LiverFemale3600.csv</a></li>
<li><strong>Clinical traits</strong>: <a href="ClinicalTraits.csv" class="external" target="_blank">ClinicalTraits.csv</a></li>
</ul>
<p>Download and set your working directory to the location of these files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the working directory</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># In RStudio: Session -&gt; Set Working Directory -&gt; Choose Directory</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"path/to/your/data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="data-input-and-pre-processing" class="level1">
<h1>Data Input and Pre-Processing</h1>
<p>This is the first step of any network analysis. I show here how to load typical expression data, pre-process them into a format suitable for network analysis, and clean the data by removing obvious outlier samples. The expression data is contained in the file LiverFemale3600.csv, linked above. After starting an R session, we load the requisite packages and the data, after appropriately setting the working directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load expression data</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>exp_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"LiverFemale3600.csv"</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="do">#####  Split and format data</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>feature_data <span class="ot">&lt;-</span> exp_data[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>] <span class="co"># Obtain data related to samples</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>assay_data <span class="ot">&lt;-</span> exp_data[, <span class="dv">9</span><span class="sc">:</span><span class="dv">143</span>] <span class="co"># Obtain gene expression data for samples</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Add names to samples</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(assay_data) <span class="ot">&lt;-</span> feature_data<span class="sc">$</span>substanceBXH</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># We must transpose the matrix so we have samples as columns</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>assay_data <span class="ot">&lt;-</span> <span class="fu">t</span>(assay_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s check the expression matrix is in the required format (genes x samples):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>assay_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      MMT00000044 MMT00000046 MMT00000051 MMT00000076 MMT00000080
F2_2    -1.81e-02     -0.0773     -0.0226    -0.00924     -0.0487
F2_3     6.42e-02     -0.0297      0.0617    -0.14500      0.0582
F2_14    6.44e-05      0.1120     -0.1290     0.02870     -0.0483
F2_15   -5.80e-02     -0.0589      0.0871    -0.04390     -0.0371
F2_19    4.83e-02      0.0443     -0.1150     0.00425      0.0251</code></pre>
</div>
</div>
<p>Note that the genes are referred to by the probe names (e.g.&nbsp;<code>MMT00000044</code>) (this data comes from expression microarray analysis). This is fine for now, but later will hamper the interpretation of results.</p>
</section>
<section id="identification-of-outlier-microarray-samples" class="level1">
<h1>Identification of outlier microarray samples</h1>
<p>Next we cluster the samples to see if there are any obvious outliers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Cluster Data and remove outliers</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># We are now going to calculate how similar the samples to each other</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>distance_matrix <span class="ot">&lt;-</span> <span class="fu">dist</span>(assay_data)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We will now perform hierarchical clustering with the distance matrix</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>sampleTree <span class="ot">&lt;-</span> <span class="fu">hclust</span>(distance_matrix, <span class="at">method =</span> <span class="st">"average"</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Now plot the dendogram</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sampleTree)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a horizontal line - we are going to take the biggest cluster below here</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> <span class="dv">15</span>, <span class="at">col =</span> <span class="st">"red"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="coexp_files/figure-html/outlier-detection-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>It appears there is one outlier (sample F2_221). One can remove it by hand, or use an automatic approach—we choose a height cut that will remove the offending sample, say 15, and use a branch cut at that height.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove outliers - "cut the tree"</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>clust <span class="ot">&lt;-</span> <span class="fu">cutreeStatic</span>(sampleTree, <span class="at">cutHeight =</span> <span class="dv">15</span>, <span class="at">minSize =</span> <span class="dv">10</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># How do they belong to clusters?</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(clust)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>clust
  0   1 
  1 134 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clust 1 contains the samples we want to keep - subset assay_data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>datExpr <span class="ot">&lt;-</span> assay_data[clust<span class="sc">==</span><span class="dv">1</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The variable datExpr now contains the expression data ready for network analysis.</p>
</section>
<section id="clinical-traits-integration" class="level1">
<h1>Clinical Traits Integration</h1>
<p>We now read in the trait data and match the samples for which they were measured to the expression samples. We also visualize the traits data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>clin_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"ClinicalTraits.csv"</span>, <span class="at">row.names=</span><span class="dv">2</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>clin_data <span class="ot">&lt;-</span> clin_data[<span class="fu">match</span>(<span class="fu">rownames</span>(datExpr), <span class="fu">rownames</span>(clin_data)), ] <span class="co"># Select samples and order them according to exp_data</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>clin_data <span class="ot">&lt;-</span> clin_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">15</span>)] <span class="co"># Remove mice attributes</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>datTraits <span class="ot">&lt;-</span> clin_data[, <span class="fu">c</span>(<span class="dv">10</span><span class="sc">:</span><span class="dv">35</span>) ] <span class="co"># Further selection of mice attributes</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>datTraits[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      weight_g length_cm ab_fat other_fat total_fat X100xfat_weight Trigly
F2_2      38.0      10.5   3.81      2.78      6.59       17.342105     14
F2_3      33.5      10.8   1.70      2.05      3.75       11.194030    109
F2_14     33.9      10.0   1.29      1.67      2.96        8.731563      2
F2_15     44.3      10.3   3.62      3.34      6.96       15.711061     71
F2_19     32.9       9.7   2.08      1.85      3.93       11.945289     55
F2_20     44.8      10.3   3.72      3.20      6.92       15.446429     34
F2_23     36.9      10.0   2.07      2.80      4.87       13.197832    198
F2_24     39.5      10.0   2.79      2.62      5.41       13.696203    186
F2_26     42.5       9.9   2.68      4.56      7.24       17.035294    357
F2_37     25.7       9.4   0.42      0.60      1.02        3.968872    109
      Total_Chol HDL_Chol  UC
F2_2        1646       34 668
F2_3        1216       27 402
F2_14        834       17 354
F2_15       1565       41 536
F2_19       1060       41 411
F2_20       1172       39 448
F2_23       1530       36 636
F2_24       1434       23 517
F2_26       1758       56 644
F2_37       1305       23 443</code></pre>
</div>
</div>
<p>We now have the expression data in the variable datExpr, and the corresponding clinical traits in the variable datTraits. Before we continue with network construction and module detection, we visualize how the clinical traits relate to the sample dendrogram.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sampleTree2 <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(datExpr), <span class="at">method =</span> <span class="st">"average"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>traitColors <span class="ot">&lt;-</span> <span class="fu">numbers2colors</span>(datTraits, <span class="at">signed =</span> <span class="cn">FALSE</span>) <span class="co"># Convert traits to a color representation: white means low, red means high, grey means missing entry</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDendroAndColors</span>(sampleTree2, traitColors,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">groupLabels =</span> <span class="fu">names</span>(datTraits), </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">main =</span> <span class="st">"Sample dendrogram and trait heatmap"</span>)   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="coexp_files/figure-html/sample-traits-dendro-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In the plot, white means a low value, red a high value, and grey a missing entry.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load clinical traits</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>traitData <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"ClinicalTraits.csv"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Match traits to expression samples</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>allTraits <span class="ot">&lt;-</span> traitData[, <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">11</span><span class="sc">:</span><span class="dv">36</span>)]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>traitRows <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">rownames</span>(datExpr), allTraits<span class="sc">$</span>Mice)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>datTraits <span class="ot">&lt;-</span> allTraits[traitRows, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(datTraits) <span class="ot">&lt;-</span> allTraits[traitRows, <span class="dv">1</span>]</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize sample dendrogram with traits</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>sampleTree2 <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(datExpr), <span class="at">method =</span> <span class="st">"average"</span>)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>traitColors <span class="ot">&lt;-</span> <span class="fu">numbers2colors</span>(datTraits, <span class="at">signed =</span> <span class="cn">FALSE</span>)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDendroAndColors</span>(sampleTree2, traitColors, <span class="at">groupLabels =</span> <span class="fu">names</span>(datTraits),</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">main =</span> <span class="st">"Sample dendrogram and trait heatmap"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p>Analysis of network topology to choose the soft thresholding power</p>
<p>Constructing a weighted gene network entails the choice of the soft thresholding power <span class="math inline">\(\beta\)</span> to which co-expression similarity is raised to calculate adjacency [1]. The authors of [1] have proposed to choose the soft thresholding power based on the criterion of approximate scale-free topology. Here we will use the function pickSoftThreshold that performs the analysis of network topology and aids the user in choosing a proper soft-thresholding power. The user chooses a set of candidate powers, and the function returns a set of network indices that should be inspected, for example as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>powers <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>), <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">12</span>, <span class="at">to=</span><span class="dv">20</span>, <span class="at">by=</span><span class="dv">2</span>))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Call the network topology analysis function and try these values</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>sft <span class="ot">&lt;-</span> <span class="fu">pickSoftThreshold</span>(datExpr, <span class="at">powerVector =</span> powers, <span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>pickSoftThreshold: will use block size 3600.
 pickSoftThreshold: calculating connectivity for given powers...
   ..working on genes 1 through 3600 of 3600</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: executing %dopar% sequentially: no parallel backend registered</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Power SFT.R.sq  slope truncated.R.sq mean.k. median.k. max.k.
1      1   0.0278  0.345          0.456  747.00  762.0000 1210.0
2      2   0.1260 -0.597          0.843  254.00  251.0000  574.0
3      3   0.3400 -1.030          0.972  111.00  102.0000  324.0
4      4   0.5060 -1.420          0.973   56.50   47.2000  202.0
5      5   0.6810 -1.720          0.940   32.20   25.1000  134.0
6      6   0.9020 -1.500          0.962   19.90   14.5000   94.8
7      7   0.9210 -1.670          0.917   13.20    8.6800   84.1
8      8   0.9040 -1.720          0.876    9.25    5.3900   76.3
9      9   0.8590 -1.700          0.836    6.80    3.5600   70.5
10    10   0.8330 -1.660          0.831    5.19    2.3800   65.8
11    12   0.8530 -1.480          0.911    3.33    1.1500   58.1
12    14   0.8760 -1.380          0.949    2.35    0.5740   51.9
13    16   0.9070 -1.300          0.970    1.77    0.3090   46.8
14    18   0.9120 -1.240          0.973    1.39    0.1670   42.5
15    20   0.9310 -1.210          0.977    1.14    0.0951   38.7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the results:</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># 1) Scale-free topology fit index as a function of the soft-thresholding power</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># We want the minimal power where the resulting network has a scale-free property.</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sft<span class="sc">$</span>fitIndices[,<span class="dv">1</span>], <span class="sc">-</span><span class="fu">sign</span>(sft<span class="sc">$</span>fitIndices[,<span class="dv">3</span>])<span class="sc">*</span>sft<span class="sc">$</span>fitIndices[,<span class="dv">2</span>],</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"Threshold"</span>,<span class="at">ylab=</span><span class="st">"Scale Free Topology Model Fit"</span>,</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="co"># this line corresponds to using an model filt cut-off of 0.9</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fl">0.90</span>,<span class="at">col=</span><span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="coexp_files/figure-html/pick-theshold-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can also see how the value affects average connectivity</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sft<span class="sc">$</span>fitIndices[,<span class="dv">1</span>], sft<span class="sc">$</span>fitIndices[,<span class="dv">5</span>],</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"Threshold"</span>,<span class="at">ylab=</span><span class="st">"Mean Connectivity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="coexp_files/figure-html/pick-theshold-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The user needs to choose the lowest power for which the scale-free topology fit index curve flattens out upon reaching a high value (in this case, roughly 0.90).</p>
<p>Constructing the gene network and identifying modules is now a simple function call. Remember one must choose the correct value for <span class="math inline">\(\beta\)</span> based on the analysis of the previous plot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pow <span class="ot">&lt;-</span> <span class="dv">6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="network-construction-and-module-detection" class="level1">
<h1>Network Construction and Module Detection</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This function implements the core of WGCNA - cluster genes and assign to modules</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>net <span class="ot">&lt;-</span> <span class="fu">blockwiseModules</span>(datExpr, <span class="at">power =</span> pow,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">TOMType =</span> <span class="st">"unsigned"</span>, <span class="at">minModuleSize =</span> <span class="dv">30</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">reassignThreshold =</span> <span class="dv">0</span>, <span class="at">mergeCutHeight =</span> <span class="fl">0.25</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">numericLabels =</span> <span class="cn">TRUE</span>, <span class="at">pamRespectsDendro =</span> <span class="cn">FALSE</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">saveTOMs =</span> <span class="cn">TRUE</span>,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">saveTOMFileBase =</span> <span class="st">"femaleMouseTOM"</span>, </span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">verbose =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Calculating module eigengenes block-wise from all genes
   Flagging genes and samples with too many missing values...
    ..step 1
Cluster size 3600 broken into 2108 1492 
Cluster size 2108 broken into 1126 982 
Done cluster 1126 
Done cluster 982 
Done cluster 2108 
Done cluster 1492 
 ..Working on block 1 .
    TOM calculation: adjacency..
    ..will not use multithreading.
     Fraction of slow calculations: 0.396405
    ..connectivity..
    ..matrix multiplication (system BLAS)..
    ..normalization..
    ..done.
   ..saving TOM for block 1 into file femaleMouseTOM-block.1.RData
 ....clustering..
 ....detecting modules..
 ....calculating module eigengenes..
 ....checking kME in modules..
     ..removing 1 genes from module 1 because their KME is too low.
     ..removing 1 genes from module 7 because their KME is too low.
     ..removing 1 genes from module 8 because their KME is too low.
     ..removing 1 genes from module 21 because their KME is too low.
 ..merging modules that are too close..
     mergeCloseModules: Merging modules whose distance is less than 0.25
       Calculating new MEs...</code></pre>
</div>
</div>
<p>We have chosen a relatively large minimum module size of 30, and a medium sensitivity deepSplit=2 for cluster splitting. The parameter mergeCutHeight is the threshold for merging of modules. We have also instructed the function to return numeric, rather than color, labels for modules, and to save the Topological Overlap Matrix. The output of the function may seem somewhat cryptic, but it is easy to use. For example, <code>net$colors</code> contains the module assignment, and net$MEs contains the module eigengenes of the modules.</p>
<p><strong>Note: The module eigengene ME is defined as the first principal component of a given module. It can be considered as a representative profile for a given co-expression module, that summarizes the expression patterns of all genes in the module into a single vector.</strong></p>
<div class="callout callout-style-default callout-tip callout-titled" title="But what exactly is an eigengene?">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
But what exactly is an eigengene?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Module eigengenes are summary statistics that capture the “gene expression profile” of a particular module in a gene co-expression network (like the ones used in WGCNA). They serve as a representative measure for the overall activity of the genes within a module, simplifying the analysis by reducing the dimensionality of the gene expression data.</p>
<section id="how-module-eigengenes-are-calculated" class="level3">
<h3 class="anchored" data-anchor-id="how-module-eigengenes-are-calculated">How Module Eigengenes Are Calculated:</h3>
<ol type="1">
<li><strong>Create a Gene Co-expression Network:</strong>
<ul>
<li>Genes are grouped into modules based on their co-expression (i.e., genes that show similar expression patterns across samples are clustered together).</li>
</ul></li>
<li><strong>Principal Component Analysis (PCA):</strong>
<ul>
<li>For each module, you perform <strong>PCA</strong> on the gene expression data of the genes within that module. The first principal component (PC1) from this analysis captures the direction of maximum variance (i.e., the most important variation in the expression profiles of the genes within the module).</li>
</ul></li>
<li><strong>Module Eigengene:</strong>
<ul>
<li>The first principal component (PC1) of the module is used as the <strong>module eigengene</strong>. This eigengene represents the overall expression pattern for that module, which can be interpreted as a single “summary” or “profile” of the genes in the module.</li>
</ul></li>
</ol>
</section>
<section id="why-module-eigengenes-are-useful" class="level3">
<h3 class="anchored" data-anchor-id="why-module-eigengenes-are-useful">Why Module Eigengenes Are Useful:</h3>
<ul>
<li><strong>Dimensionality Reduction:</strong> Instead of analyzing each gene individually, module eigengenes provide a summary for each gene module, reducing the complexity of the data.</li>
<li><strong>Biological Interpretation:</strong> Module eigengenes represent the combined expression pattern of all genes in a module, making it easier to study the overall behavior of a biological process, pathway, or condition related to that module.</li>
<li><strong>Correlation with Traits:</strong> Module eigengenes can be correlated with external traits (e.g., disease status, clinical measures), helping identify modules that are associated with specific biological conditions or phenotypes.</li>
</ul>
</section>
<section id="example" class="level3">
<h3 class="anchored" data-anchor-id="example">Example:</h3>
<p>If you have a module of genes related to immune response, the module eigengene will summarize the expression pattern of all those immune-related genes across samples. A high value of the module eigengene in a certain condition (e.g., cancer) might indicate that the immune-related genes in the module are highly active in that condition.</p>
<p>Thus, module eigengenes are a powerful tool for simplifying the analysis of large-scale gene co-expression networks and understanding the underlying biological processes.</p>
</section>
</div>
</div>
</div>
<p>To see how many modules were identified and what the module sizes are, one can use: table(net$colors)</p>
<p>Its output indicates that there are 18 modules, labeled 1 through 18 in order of descending size, with sizes ranging from 609 to 34 genes. The label 0 is reserved for genes outside of all modules. The hierarchical clustering dendrogram (tree) used for the module identification is returned in <code>net$dendrograms[[1]]</code>. The dendrogram can be displayed together with the color assignment using the following code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>moduleColors <span class="ot">&lt;-</span> <span class="fu">labels2colors</span>(net<span class="sc">$</span>colors)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot function</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotDendroAndColors</span>(net<span class="sc">$</span>dendrograms[[<span class="dv">1</span>]], moduleColors[net<span class="sc">$</span>blockGenes[[<span class="dv">1</span>]]],</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"Module colours"</span>,</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">dendroLabels =</span> <span class="cn">FALSE</span>, <span class="at">hang =</span> <span class="fl">0.03</span>,</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">addGuide =</span> <span class="cn">TRUE</span>, <span class="at">guideHang =</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="coexp_files/figure-html/view-module-clustering-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Here we see the genes clustered into different modules, corresponding to the different colours in the module colours band.</p>
</section>
<section id="correlating-modules-with-external-clinical-traits" class="level1">
<h1>Correlating modules with external clinical traits</h1>
<section id="quantifying-moduletrait-associations" class="level2">
<h2 class="anchored" data-anchor-id="quantifying-moduletrait-associations">Quantifying module–trait associations</h2>
<p>We would like to identify modules that are significantly associated with the measured clinical traits. Since we already have a summary profile (eigengene) for each module, we simply correlate eigengenes with external traits and look for the most significant associations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define numbers of genes and samples</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>nSamples <span class="ot">&lt;-</span> <span class="fu">nrow</span>(datExpr)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate module eigen gene values for each module - </span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Effectively the average expression of all genes in the module accross samples</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>MEs0 <span class="ot">&lt;-</span> <span class="fu">moduleEigengenes</span>(datExpr, moduleColors)<span class="sc">$</span>eigengenes</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>MEs <span class="ot">&lt;-</span> <span class="fu">orderMEs</span>(MEs0)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Correlate the modules with the traits</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>moduleTraitCor <span class="ot">&lt;-</span> <span class="fu">cor</span>(MEs, datTraits, <span class="at">use =</span> <span class="st">"p"</span>)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>moduleTraitPvalue <span class="ot">&lt;-</span> <span class="fu">corPvalueStudent</span>(moduleTraitCor, nSamples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-module-trait-correlations" class="level2">
<h2 class="anchored" data-anchor-id="plot-module-trait-correlations">Plot module-trait correlations</h2>
<p>Since we have a moderately large number of modules and traits, a suitable graphical representation will help in reading the table. We color code each association by the correlation value:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is to set parameters for plotting</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">6</span>, <span class="fl">8.5</span>, <span class="dv">3</span>, <span class="dv">3</span>))</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the correlation values within a heatmap plot</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">labeledHeatmap</span>(<span class="at">Matrix =</span> moduleTraitCor,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">xLabels =</span> <span class="fu">names</span>(datTraits),</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>               <span class="at">yLabels =</span> <span class="fu">names</span>(MEs),</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">ySymbols =</span> <span class="fu">names</span>(MEs),</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">colorLabels =</span> <span class="cn">FALSE</span>,</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>               <span class="at">colors =</span> <span class="fu">blueWhiteRed</span>(<span class="dv">50</span>),</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>               <span class="at">setStdMargins =</span> <span class="cn">FALSE</span>,</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>               <span class="at">cex.text =</span> <span class="fl">0.5</span>,</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>               <span class="at">zlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>),</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>               <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"Module-trait relationships"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="coexp_files/figure-html/plot-cor-eigen-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The analysis identifies the several interesting module–trait associations. We will now focus on weight as the trait of interest. Note that for example the brown module represents a group of genes that is correlated with weight, abdominal and other fat levels.</p>
</section>
</section>
<section id="gene-relationship-to-trait-and-important-modules-gene-significance-and-module-membership" class="level1">
<h1>Gene relationship to trait and important modules: Gene Significance and Module Membership</h1>
<p>We quantify associations of individual genes with our trait of interest (weight) by defining Gene Significance GS as (the absolute value of) the correlation between the gene and the trait. For each module, we also define a quantitative measure of module membership MM as the correlation of the module eigengene and the gene expression profile. This allows us to quantify the similarity of all genes on the array to every module. Mathematically, for a gene ( g ) in a module ( M ): <span class="math display">\[
\text{MM}(g) = \text{cor}(\text{Expr}(g), \text{ME}(M)),
\]</span> where: - <span class="math inline">\(\text{Expr}(g)\)</span>: Gene <span class="math inline">\(g\)</span>’s expression profile across all samples. - <span class="math inline">\(\text{ME}(M)\)</span>: The module eigengene of module <span class="math inline">\(M\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Some genes in each module are more representative than others of the eigen-gene values.</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is calculated as correlation betwen module eigen genes and gene expression</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># This can be used to filter large modules to obtain the most important ones using following code:</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>geneModuleMembership <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cor</span>(datExpr, MEs, <span class="at">use =</span> <span class="st">"p"</span>))</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>MMPvalue <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">corPvalueStudent</span>(<span class="fu">as.matrix</span>(geneModuleMembership), nSamples))</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Now rename - using the eigen gene names and replacing ME with MM</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>modNames <span class="ot">&lt;-</span> <span class="fu">substring</span>(<span class="fu">names</span>(MEs), <span class="dv">3</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(geneModuleMembership) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"MM."</span>, modNames, <span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(MMPvalue) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"p.MM."</span>, modNames, <span class="at">sep=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>MM ranges from -1 to 1: - <strong>High MM values (close to 1)</strong>: The gene is strongly correlated with the module eigengene, making it a core member. - <strong>Low MM values (close to 0)</strong>: The gene has little to no relationship with the module. - <strong>Negative MM values</strong>: The gene’s expression pattern is inversely related to the module eigengene.</p>
<section id="why-is-module-membership-useful" class="level3">
<h3 class="anchored" data-anchor-id="why-is-module-membership-useful">Why Is Module Membership Useful?</h3>
<ol type="1">
<li><strong>Prioritizing Key Genes in Modules:</strong>
<ul>
<li>Genes with high MM values are considered the “hub” or <strong>core genes</strong> of the module. These genes are most representative of the module’s behavior and are likely to play central roles in the biological processes associated with the module.</li>
</ul></li>
<li><strong>Identifying Candidate Biomarkers:</strong>
<ul>
<li>High-MM genes are strong candidates for biomarkers or therapeutic targets because they drive the module’s co-expression patterns, which may relate to specific phenotypes or traits.</li>
</ul></li>
<li><strong>Understanding Module Structure:</strong>
<ul>
<li>MM helps distinguish between:
<ul>
<li><strong>Core genes:</strong> Essential for the module’s coherence.</li>
<li><strong>Peripheral genes:</strong> Weakly connected or partially related to the module.</li>
</ul></li>
<li>This understanding is valuable for functional annotation and interpretation of the module.</li>
</ul></li>
<li><strong>Functional Enrichment Analysis:</strong>
<ul>
<li>MM values can help refine functional enrichment studies. Analyzing only high-MM genes within a module often enhances the biological signal by focusing on genes that are most relevant.</li>
</ul></li>
<li><strong>Identifying Outliers or Ambiguous Genes:</strong>
<ul>
<li>Low-MM genes in a module may indicate potential outliers, genes that don’t fit well in any module, or genes that are shared between multiple modules.</li>
</ul></li>
</ol>
</section>
</section>
<section id="summary-output-of-network-analysis-results" class="level1">
<h1>Summary output of network analysis results</h1>
<p>We have found modules with high association with our trait of interest, and have identified their central players by the Module Membership measure. We now merge this statistical information with gene annotation and write out a file that summarizes the most important results and can be inspected in standard spreadsheet software such as MS Excel or Open Office Calc. Our expression data are only annotated by probe ID names: the command will return probe IDs belonging to the brown module. To facilitate interpretation of the results, we use a probe annotation file provided by the manufacturer of the expression arrays to connect probe IDs (such as MMT00000044 like we saw earlier) to gene names and universally recognized identification numbers (Entrez codes).</p>
<p>We create a data frame holding the following information for all probes: probe ID, gene symbol, Locus Link ID (Entrez code), module color, gene significance for weight, and module membership and p-values in all modules. The modules will be ordered by their significance for weight, with the most significant ones to the left.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>annot <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"GeneAnnotation.csv"</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>probes <span class="ot">&lt;-</span> <span class="fu">colnames</span>(datExpr)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># We use this to match the genes in our experiment with the annotation file column </span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>probes2annot <span class="ot">&lt;-</span> <span class="fu">match</span>(probes, annot<span class="sc">$</span>substanceBXH)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the starting data frame</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>geneInfo0 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">substanceBXH =</span> probes,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">geneSymbol =</span> annot<span class="sc">$</span>gene_symbol[probes2annot],</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">LocusLinkID =</span> annot<span class="sc">$</span>LocusLinkID[probes2annot],</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>                        <span class="at">moduleColor =</span> moduleColors)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we add to this the MM data using cbind</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>geneInfo <span class="ot">&lt;-</span> <span class="fu">cbind</span>(geneInfo0, geneModuleMembership, MMPvalue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="now-your-task---select-all-genes-that-belong-to-the-brown-module-and-perform-ora-to-see-if-you-can-find-enrichment-for-any" class="level1">
<h1>Now your task - select all genes that belong to the brown module and perform ORA to see if you can find enrichment for any</h1>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>